### ISSUE TITLE
How to handle providers/processors when headless

### ISSUE THREAD

**Issue by lucyleeow** [Napari/napari#6917]
This is a request for comments about how to handle app-model actions in headless mode (i.e. without a Qt `Viewer`). There have been discussions here and there but nothing well documented and no concrete decisions. Recent PRs have brought up this topic and this is an attempt at gaining more clarity.

Originally, there was separation of actions/providers/processors (actions for short) into 'GUI/Qt' and 'non-GUI/Qt'. But it seems that most of the actions are technically Qt (see #6767, #6883, #6743)

Considerations about how to separate 'GUI/Qt' and 'non-GUI/Qt' action files:

* We think that actions that do not technically require QT but are defined for the sole purpose of living in a menu, should be considered a 'GUI' action. (ref: https://github.com/napari/napari/pull/6848#issuecomment-2089374696) - e.g., help menu actions that open up a URL.
* Small factor - it's easier to have all actions belonging to a menu (e.g., file menu) in one file, so they are easier to find
* Having all actions in one file makes it easier to include submenu definitions with the actions, as we don't need to choose which file to define the submenus in (ref: #6848)

The aim of separating 'Qt' and 'non-Qt' actions is such that non-Qt actions are registered and available in the `app` headless. But this brings more considerations:

1. Which built-in actions should be made available to the headless app?
    * I think all (?) the menubar actions that might be useful headless, already have a public API function (e.g., `open_sample`)
   * Some layer actions would make sense headless but see 2.
3. Currently there are no 'non-Qt' providers. This means that the layer actions which technically do not need Qt, won't work without Qt (the layerlist provider uses `_viewer_provider` which gets the active viewer via `_QtMainWindow`)
    * when does headless dependency injection make sense? When headless we do not know which viewer/layer is active (discussed more here #6768).
    * if headless does not use dependency injection, currently the user would need to manually pass the argument to the action. Do we want to add a provider that guesses which e.g., layer the user wants to perform the action on? 

**What do we want headless action API to look like?**

* Registering actions - what should the API look like? Should we let users directly use `app.register_action` or should we add a wrapper API? Headless and GUI users should both be able to do this (GUI, via console)
* Executing actions - done via ``CommandRegistry.execute_command(<command_id_str>)`` or different API?

cc @DragaDoncila @jni @psobolewskiPhD @Czaki who may be interested. Thank you

**Comment by DragaDoncila** [comment]
> We think that actions that do not technically require QT but are defined for the sole purpose of living in a menu, should be considered a 'GUI' action.

I think this bit here is very tricky because it's a a consideration on how people want to use napari. For example, take any given `Open sample` action that currently lives in the file menu. This is obviously not only a GUI action - people want to open files in headless mode too! However, there's already a blessed API for that that has nothing to do with the `app-model` file menu open sample action. Would a user ever want to trigger the menu version of the open sample action in headless mode? I think the answer is no. Does that mean the file menu actions for opening samples are GUI actions? Lucy and I have basically decided yes. This also means the action is not registered until all other GUI actions are registered. We think this makes sense - we don't want actions registered with the app that are guaranteed to fail right?

However, there is then the further distinction of GUI vs. Qt. Is the `File -> Open Sample` action **qt-specific**? No, probably not. Any frontend we used would likely include some way of defining menubar actions. Where should this action live then, structurally speaking, given that we definitely don't want it to be registered and available for execution in headless mode, but it's also not specifically Qt dependent?

On the topic of dependency injection in headless mode, I have no real idea whether it makes sense or not, or what parameters users might request that they can't explicitly pass in but will be available at runtime. The current providers for viewer and layers work on the basis that through the GUI, you can infer which viewer/layer the user wanted associated with a particular action e.g. because a certain viewer has focus or because a certain layer is active and selected in the layerlist. Should we (and can we even) make those same inferences in headless mode? It doesn't seem natural...

**Comment by lucyleeow** [comment]
> For example, take any given Open sample action that currently lives in the file menu.

I think that falls under:

>  I think all (?) the menubar actions that might be useful headless, already have a public API function (e.g., open_sample)

But I should have been more specific in the kind of actions I was talking about there. E.g., the help menu actions that open up a URL.

---

> given that we definitely don't want it to be registered and available for execution in headless mode, but it's also not specifically Qt dependent?

I think technically the open sample one is Qt dependent because we use `handle_gui_reading` which uses Qt. But this does raise a good question about functions which get a 'wrapper' for actions that get added to the menu.
In this case we have the `open_sample` API which users can use directly. Potentially there would be other actions where the menu version gets a wrapper which is not needed headless? In this case do we have 2 actions??

>  Where should this action live then, structurally speaking, given that we definitely don't want it to be registered and available for execution in headless mode, but it's also not specifically Qt dependent?

I feel like this question is out of scope of this issue. It's a long standing issue of napari, that there is not GUI but not Qt folder? It's touched upon here: https://github.com/napari/napari/issues/6469

**Comment by psobolewskiPhD** [comment]
Question: am I misunderstanding or there should be a distinction between `headless` and programmatic?
I personally don't see much value (or have interest in) the first, but the 2nd, programmatic access to things in an viewer, is something that is absolute gold and we should strive for 1:1 equivalence as much as possible.

**Comment by lucyleeow** [comment]
Sorry I think I mean headless.

i.e., `init_qactions` (which registers all the Qt actions) gets run when we init `_QtMainWindow`. 

So whenever there is a Qt `Viewer` (vs `ViewerModel`) all the Qt actions and such are registered. So they are available for use.

Note also the other question is how we want the user to trigger there actions. This is currently possible via their command ID string, but this is a bit long (and we are not exactly sure how to name it... see https://github.com/napari/docs/pull/405). E.g., for the help actions the strings are currently e.g., 'napari.window.help.info'.
I would think that a different public API to accessing such actions would be nicer.

**Comment by lucyleeow** [comment]
Found this while doing some digging:

> Something like accessors+processors gives pure Puthon functions reasonable default behavior in the napari/GUI context while leaving the door open for alternative contexts such as execution headless, cloud, cluster, or embedded in other applications. 

from: https://github.com/napari/napari/issues/4532#issuecomment-1128990549

I think I am also leaning towards processors/providers being GUI-only.

cc @DragaDoncila

**Comment by lucyleeow** [comment]
But also this gives some perspective on what plugin contributions should look like:

> The original ImageJ assumes it is a standalone GUI application, and so plugins do everything explicitly like constructing and showing dialogs and images, which are built on GUI toolkits directly. So then there is no way to use the associated algorithm code without these side effects happening too.

https://github.com/napari/napari/issues/4532#issuecomment-1128990549

**Comment by jni** [comment]
do you mean *shouldn't* look like? 😅

**Comment by lucyleeow** [comment]
From decision in #6946, it would follow that we do want to register actions that would be useful to the user and can be run without a GUI (i.e. some layer actions). Atm the user would need to pass the layer they want to run the action on, as current providers cannot do so headless.

But the question remaining is: Do we want to add a provider that guesses which e.g., layer the user wants to perform the action on?

cc @Czaki

**Comment by lucyleeow** [comment]
Looking at the other actions, I think we have 3 types:

* requires Qt
* requires GUI, not necessarily Qt
* can be run headless, user would need to pass in dependency

I agree that the latter should be available in the `app` even headless (i.e. the layer actions https://github.com/napari/napari/issues/6946). The first 2 have been lumped together as we don't have a GUI not not necessarily Qt folder, and I think is fine for now. This also makes sense as we don't.

The only thing that this brings to question are the help menu actions, which just open a URL. These can be used headless but live in Qt (as they are defined solely for purpose of living in a menu). I think I am, because they aren't really useful.

The main question really is how providers should work. Should we ask the user to pass dependency or should we add a 'headless' provider that guesses?

**Comment by jni** [comment]
> The main question really is how providers should work. Should we ask the user to pass dependency or should we add a 'headless' provider that guesses?

Personally I like guessing *only* when it's very clear how one would guess (e.g. selected layer). Also, there should be some sort of escape hatch for guessing. For example, in the context of a menu, you could hold option before click to get the full selection widget. In the context of a keyboard shortcut, you'd have two shortcuts, one for the "guess" action and one for the "ask" action (ideally, these are very similar, eg. Cmd-T vs Cmd-Option-T).

**Comment by lucyleeow** [comment]
> Personally I like guessing only when it's very clear how one would guess

Yes i agree but last we talked you suggested we could try guessing in headless mode (it was at end of a meeting) !

The providers we have atm are pretty simple, active viewer/qt viewer/window/layer/layerlist.

**Comment by DragaDoncila** [comment]
In `headless` mode I think the user should always be explicitly passing the dependency. I know the `layerlist` and its `selection` objects exist in the `headless` context, but I don't think there should be actions e.g. `segment_layer(layer: 'napari.layers.Image')`, where we pass in an image layer if it's selected, and otherwise fail. 

Because we're in `headless`, we wouldn't have the option of presenting the user with a dialog/command prompt to select which layer they want to pass, so our only valid option would be to raise an error in the case where someone tried to execute that action without passing a layer (and we didn't have an image layer selected). Raising in this situation might be super annoying for the user, as they might be batch executing some commands, which might even change the layerlist selection (or add new layers which would trigger a change in the layerlist selection afaik?), and then fail halfway through on a command that previously worked just fine.

TLDR: in headless mode, it's the user's job to satisfy all dependencies of actions they are executing, by explicitly passing in the correct function parameters.

**Comment by lucyleeow** [comment]
Thanks @DragaDoncila I totally agree.

I think maybe @jni was referring to use in GUI mode when he said:

> Also, there should be some sort of escape hatch for guessing. For example, in the context of a menu, you could hold option before click to get the full selection widget. In the context of a keyboard shortcut, you'd have two shortcuts, one for the "guess" action and one for the "ask" action (ideally, these are very similar, eg. Cmd-T vs Cmd-Option-T).

I'm also in agreement that the user should always pass dependency in headless mode.

I am happy to close this, but do you think we should try to solicit more opinions?

**Comment by psobolewskiPhD** [comment]
I can't really wrap my brain around what a headless napari is/would be and what I would use it for, so I fear I'm largely useless on this topic, but Draga's answer seems the most cogent to me? 🤷

**Comment by DragaDoncila** [comment]
> I can't really wrap my brain around what a headless napari is/would be and what I would use it for

I have no idea where the `headless` name came from, or even if I'm using it the way it's supposed to be used 😅 but as far as I know, the `headless` `Viewer` is a `Viewer` object that is created without its GUI i.e. the `Viewer` object you have a handle on before you hit `napari.run` or when you don't have `qt` installed... @jni would you say that's accurate...?

**Comment by ctrueden** [comment]
> I have no idea where the `headless` name came from

Chiming in late here, since no one else talked about this: "headless" is a technical term meaning an execution environment *without a [display](https://en.wikipedia.org/wiki/Electronic_visual_display)* (or even necessarily a graphics card). See [Headless software](https://en.wikipedia.org/wiki/Headless_software) article on Wikipedia.

Can you run Python on a text-only server only accessible via SSH? The answer is almost certainly yes.

Can you run *napari* on such a computer? The answer is almost certainly no, because one of the first things the napari program does is spin up a Qt event loop, which is going to try talking to a display, which is going to crash or at least raise a fatal exception.

> I can't really wrap my brain around what a headless napari is/would be

It would mean the ability to use napari *data structures* without using the napari *GUI*. Layer objects, for example, have important metadata that one might want to work with, even without actually displaying anything. Furthermore: one might want to do so-called "off-screen rendering" (sadly, no Wikipedia article about this concept that I could find, even though articles like [Pixel buffer](https://en.wikipedia.org/wiki/Pixel_buffer) and [Headless browser](https://en.wikipedia.org/wiki/Headless_browser) use the term without defining it)—generating visualizations and serializing them to disk in some image format like PNG.

Given napari's commitment to rendering performance, it is presumably rather inextricably tied to hardware-accelerated rendering with GPUs, making true off-screen rendering operations less tenable. But these days, when someone says "headless" they often/usually mean only "without a screen" or "without a [windowing system](https://en.wikipedia.org/wiki/Windowing_system) running." So napari could keep its more efficient GPU-driven rendering pipeline, but perform [headless rendering](https://en.wikipedia.org/wiki/Headless_software#Headless_rendering), utilizing the rendering backend's ability to push pixels into an off-screen [framebuffer](https://en.wikipedia.org/wiki/Framebuffer) (e.g. OpenGL and Vulkan both support doing this).

**Comment by jni** [comment]
> (sadly, no Wikipedia article about [off-screen rendering] that I could find, even though articles like [Pixel buffer](https://en.wikipedia.org/wiki/Pixel_buffer) and [Headless browser](https://en.wikipedia.org/wiki/Headless_browser) use the term without defining it)

Seems like an opportunity to contribute to humanity's collective knowledge @ctrueden 😃
